codes/c/chapter_backtracking/permutations_i.c

```diff
@@ -0,0 +1,94 @@
+/**
+ * File: permutations_i.c
+ * Created Time: 2023-06-04
+ * Author: Gonglja (glj0@outlook.com)
+ */
+
+#include "../utils/common.h"
+
+/* 回溯算法：全排列 I */
+void backtrack(vector *state, vector *choices, vector *selected, vector *res) {
+    // 当状态长度等于元素数量时，记录解
+    if (state->size == choices->size) {
+        vector *newState = newVector();
+        for (int i = 0; i < state->size; i++) {
+            vectorPushback(newState, state->data[i]);
+        }
+        vectorPushback(res, newState);
+        return;
+    }
+    // 遍历所有选择
+    for (int i = 0; i < choices->size; i++) {
+        int *choice = malloc(sizeof(int));
+        *choice = *((int *)(choices->data[i]));
+        // 剪枝：不允许重复选择元素 且 不允许重复选择相等元素
+        bool select = *((bool *)(selected->data[i]));
+        if (!select) {
+            // 尝试：做出选择，更新状态
+            *((bool *)selected->data[i]) = true;
+            vectorPushback(state, choice);
+            // 进行下一轮选择
+            backtrack(state, choices, selected, res);
+            // 回退：撤销选择，恢复到之前的状态
+            *((bool *)selected->data[i]) = false;
+            vectorPopback(state);
+        }
+    }
+}
+
+// 打印二维向量中的元素
+void printVectorMatrix(vector *vv) {
```

Comment:
Is this a tool function? Please extract it to the public module like `codes/cpp/utils/print_utils.hpp`